File: src-tauri/src/community/provider.rs
use std::{any::Any, collections::HashMap};

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use tauri::ipc::Channel;

pub mod jsplugin;
pub mod official;
pub mod bandbbs;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Author {
    pub name: Option<String>,
    pub author_url: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Item {
    pub name: Option<String>,
    pub restype: Option<String>,
    pub description: Option<String>,
    pub preview: Option<Vec<String>>,
    pub icon: Option<String>,
    pub source_url: Option<String>,
    pub author: Option<Vec<Author>>,
    #[serde(default)]
    pub paid_type: String,
    pub _bandbbs_ext_supported_device: Option<String>,
    pub _bandbbs_ext_resource_id: Option<u32>,
    pub _bandbbs_ext_is_community_paid: Option<bool>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct SearchConfig {
    pub filter: Option<String>,
    pub sort: Option<String>,
    pub category: Option<Vec<String>>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ProgressData {
    pub progress: f32,
    pub status: String,
    pub status_text: String
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum ProviderState {
    Ready,
    Updating,
    Failed(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceManifestV1 {
    pub item: Item,
    pub downloads: HashMap<String, ResDownloadInfoV1>
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResDownloadInfoV1 {
    pub version: String,
    pub file_name: String
}

#[async_trait]
pub trait Provider: Send + Sync {
    fn provider_name(&self) -> String;
    fn as_any(&self) -> &dyn Any;

    async fn refresh(&self) -> anyhow::Result<()> {
        Ok(())
    }

    fn state(&self) -> ProviderState {
        ProviderState::Ready
    }

    async fn get_page(
        &self,
        page: u32,
        limit: u32,
        search: SearchConfig,
    ) -> anyhow::Result<Vec<Item>>;

    async fn get_categories(&self) -> anyhow::Result<Vec<String>>;

    async fn get_item(&self, name: String) -> anyhow::Result<ResourceManifestV1>;

    async fn download(&self, name: String, device: String, progress_cb: Channel<ProgressData>) -> anyhow::Result<String>;

    async fn get_total_items(&self, filter: Option<String>) -> anyhow::Result<u64>;
}

File: src/community/provider.ts
import logger from "@/log/logger";
import { ProgressData, Provider, ProviderState, SearchConfig } from "@/plugin/types";
import { Item, ResourceManifestV1 } from "@/types/ResManifestV1";
import { Channel, invoke } from "@tauri-apps/api/core";

export default class BackendProvider extends Provider {
    constructor(
        public readonly name: string,
        private onRefreshComplete: () => void
    ) {
        super();
    }

    async refresh(): Promise<void> {
        try {
            await invoke("commprov_refresh", { name: this.name });
            this.onRefreshComplete();
        } catch (error) {
            console.error(`Failed to refresh provider ${this.name}:`, error);
            this.onRefreshComplete();
            throw error;
        }
    }

    async getState(): Promise<ProviderState> {
        return await invoke("commprov_get_state", { name: this.name });
    }

    async getCategories(): Promise<string[]> {
        return await invoke("commprov_get_categories", { name: this.name });
    }

    async getPage(page: number, limit: number, search: SearchConfig): Promise<Item[]> {
        return await invoke("commprov_get_page", {
            name: this.name,
            page,
            limit,
            search,
        });
    }

    async getItem(resname: string): Promise<ResourceManifestV1> {
        logger.info(`provider ${this.name} try to getItem ${resname}`)
        return await invoke("commprov_get_item", {
            name: this.name,
            resname,
        });
    }

    async download(resname: string, device: string, progressCb: Channel<ProgressData>): Promise<string> {
        return await invoke("commprov_download", {
            name: this.name,
            resname,
            device,
            progressCb,
        });
    }

    async getTotalItems(filter?: string): Promise<number> {
        return await invoke("commprov_get_total_items", {
            name: this.name,
            filter,
        });
    }
}